diff --git a/include/scsi/scsi.h b/include/scsi/scsi.h
old mode 100644
new mode 100755
index d0a66aa..7b8db5c
--- a/include/scsi/scsi.h
+++ b/include/scsi/scsi.h
@@ -198,6 +198,10 @@ enum scsi_timeouts {
 /* Vendor specific CDBs start here */
 #define VENDOR_SPECIFIC_CDB 0xc0
 
+/* Op code for ZBC */
+#define RESET_WP        0x94
+#define REPORT_ZONES    0x95
+
 /*
  *	SCSI command lengths
  */
@@ -585,6 +589,13 @@ static inline int scsi_is_wlun(u64 lun)
 /* Used to obtain the PCI location of a device */
 #define SCSI_IOCTL_GET_PCI		0x5387
 
+/* Used for Zone based SMR devices */
+#define SCSI_IOCTL_INQUIRY		0x10000
+#define SCSI_IOCTL_RESET_WP		0x10001
+#define SCSI_IOCTL_REPORT_ZONES		0x10002
+#define SCSI_IOCTL_OPEN_ZONE		0x10003
+#define SCSI_IOCTL_CLOSE_ZONE		0x10004
+
 /* Pull a u32 out of a SCSI message (using BE SCSI conventions) */
 static inline __u32 scsi_to_u32(__u8 *ptr)
 {
